[STEP 4] ♠️블랙잭(베팅) 미션 제출합니다!

y2gcoder opened the pull request:
안녕하세요 한솔님! 

블랙잭(베팅) 미션 구현하여 제출합니다 :) 

뭔가 기존 구조를 유지한 채 변경하려고 하다보니 더 어렵게 생각하나 싶습니다

한솔님의 피드백을 기다리면서 코틀린 인 액션도 읽어보고 있겠습니다!

항상 감사합니다 ☺️

양영근 올림

y2gcoder committed:
feature(blackjack): dealer의 turn 진행을 player turn 과 같이 처리해 추상화 층위 맞추기

한솔님 피드백 반영!
추후 책임을 이관할 수 있는 곳을 찾아보자!

y2gcoder committed:
docs(blackjack): 블랙잭(베팅) 구현 기능 목록 정의

y2gcoder committed:
feature(blackjack): 은행은 모든 참가자의 베팅 금액을 받는다

y2gcoder committed:
feature(blackjack): 플레이어 베팅 금액은 1원 이상이 되도록 유효섬 검사 추가

y2gcoder committed:
feature(blackjack): 은행은 각 참가자의 잔고를 확인할 수 있다

y2gcoder committed:
feature(blackjack): 은행의 정산 처리 메서드 추가 및 플레이어 점수 합 승리 정산 처리

TODO: Map 형태 리팩토링 필요

y2gcoder committed:
feature(blackjack): 은행 정산 시 딜러의 손패 합 승리 케이스 추가

y2gcoder committed:
feature(blackjack): Hand 에 블랙잭 판단 메서드 추가 && Participant 에서 블랙잭 판단 위임 메서드 추가

y2gcoder committed:
feature(blackjack): DefaultBlackJackJudgeOutcomeStrategy 에서 딜러만 블랙잭일 때의 플레이어 패배 판단 로직 추가

y2gcoder committed:
feature(blackjack): DefaultBlackJackJudgeOutcomeStrategy 에서 플레이어의 블랙잭 승리 케이스 추가

y2gcoder committed:
feature(blackjack): 게임 심판은 PUSH(딜러와 플레이어의 손 패가 같음 혹은 둘다 블랙잭) 선언을 할 수 있다

y2gcoder committed:
feature(blackjack): 은행에 플레이어가 버스트 했을 때의 정산처리 추가

y2gcoder committed:
test(blackjack): 은행에 딜러 버스트 케이스의 정산 테스트 케이스 추가

y2gcoder committed:
feature(blackjack): 은행에 플레이어 블랙잭 승리 케이스에 대한 정산 추가

y2gcoder committed:
test(blackjack): 은행에 딜러 블랙잭 승리(플레이어 패배)에 대한 정산 케이스 추가

y2gcoder committed:
test(blackjack): 은행에 딜러 블랙잭 승리(플레이어 패배)에 대한 정산 케이스 추가

y2gcoder committed:
refactor(blackjack): Bank의 settleBets 의 라인 수 및 indent depth 를 줄이기 위해 메서드 추출

y2gcoder committed:
feature(blackjack): 은행은 각 참가자의 수익률을 도출할 수 있도록 profits() 추가

y2gcoder committed:
feature(blackjack): 은행에서 GameResult 목록을 받아 정산 처리할 수 있는 settleBets() 메서드 추가

y2gcoder committed:
feature(blackjack): 베팅에 대한 UI 적용 && 필요없어진 딜러에 대한 요약 결과 발표 로직 삭제

y2gcoder committed:
test(blackjack): Bank의 예외 상황에 대한 테스트 케이스 추가

catsbi commented on src/main/kotlin/blackjack/Bank.kt:
```diff
@@ -0,0 +1,116 @@
+            require(betAmount > 0) {
+                "플레이어는 1원 이상 베팅해야 합니다 betAmount: $betAmount"
+            }
+        }
```
이렇게되면 추상화를 했지만 구현체를 알아야하는 문제가 생길 것 같아요. 추상화를 활용하여 구현체를 의존하는 경우가 없도록 리팩토링 해볼 수 없을까요?

catsbi commented on src/main/kotlin/blackjack/Bank.kt:
```diff
@@ -0,0 +1,116 @@
+
+    fun bet(
+        participant: Participant,
+        betAmount: Double = 0.0,
```
금액을 책임지는 일급 객체를 만들어보면 어떨까요? 
값 객체로 infix 함수 연산자들을 오버라이딩해서 코틀린의 기능을 잘 활용해보면 재밌을 것 같네요!

catsbi commented on src/main/kotlin/blackjack/Bank.kt:
```diff
@@ -0,0 +1,116 @@
+            Outcome.LOSS -> settleLoss(dealerAccount, playerAccount)
+            Outcome.BLACKJACK -> settleBlackJack(dealerAccount, playerAccount)
+            Outcome.PUSH -> Unit
+        }
```
각 케이스에 대한 핸들링 로직이 하나의 객체에서 함수로 구분되고 있다보니, 코드가 너무 길어지는 것 같은데, seald class나 책임사슬 패턴을 통해 확장성을 챙겨보면 어떨까요? 
```suggestion
        outcomHandlerFactory.findHandler(outcome).handle(dealerAccount, playerAccount)
```

catsbi commented on src/main/kotlin/blackjack/Hand.kt:
```diff
@@ -17,6 +17,8 @@ class Hand(
 
     fun isBust(): Boolean = sumOfHand() > BLACKJACK_NUMBER
 
+    fun isBlackJack(): Boolean = _cards.size == 2 && sumOfHand() == BLACKJACK_NUMBER
```
함수를 하나하나 구현하기보단 열거타입이나 종류별로 카테고라이징된 정보들을 토대로 책임을 위임하여, 행위를 추상화할수 없을까요? getCurState() 같은 함수등을통해서 현재 플레이어의 카드를 기반으로 어떤 상태인지 조회하는 함수가 있으면 어떨까싶어요 ㅎㅎ

catsbi commented on src/main/kotlin/blackjack/ui/BlackJackGame.kt:
```diff
@@ -55,6 +59,18 @@ private fun setupPlayers(
     Player(name = playerName, initialCards = initialCards)
 }
 
+private fun betAllParticipants(
```
현재 이 객체의 역할은 service인가요 controller인가요 ui인가요? SRP 에 대해 고민해보면 좋을 클래스로 보이네요!

catsbi commented on src/test/kotlin/blackjack/BankTest.kt:
```diff
@@ -0,0 +1,270 @@
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.shouldBe
+
+class BankTest : StringSpec({
```
이제 게임진행의 복잡도가 높아진만큼 StringSpec도 좋지만 BehaviorSpec과 같은 행위주도 방식의 테스트 작성을 해보는건 어떨까요? 

게임의 흐름에 따른 여러 브랜치 커버리지를 고려해볼 수 있을 것 같네요!

catsbi commented on src/test/kotlin/blackjack/BankTest.kt:
```diff
@@ -0,0 +1,270 @@
+        sut.balance(dealer) shouldBe 0
+    }
+
+    "은행은 딜러가 베팅할 때 베팅 금액을 생략하면 0.0으로 베팅한다" {
```
사실 은행이라는 개념이 갑자기 들어온게 적절한가 의문이듭니다. 🤔 
카지노 게임장을 가더라도 딜러와 플레이어간의 게임이지 중간에 은행이 있지는 않으니까요. 

지금은 마치 은행이라는 딜러와, 플레이어1(딜러), 플레이어2(게이머) 게임이 진행되는 것 같아요. ㅎㅎ

catsbi reviewed (changes requested):
영근님 어느덧 블랙잭 마지막 미션이군요! 
개념을 잘 잡고 작성을 해주셔서 크게 피드백 드릴게 없이도 잘 진행된 것 같아요.  👏 

해당 마지막 미션 몇 가지 피드백 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요!

y2gcoder committed:
feature(blackjack): Bank 내부의 베팅 금액을 Money 라는 value class 로 래핑

y2gcoder committed:
refactor(blackjack): Card 에서 필요 없는 Comparable 구현 제거

y2gcoder committed:
refactor(blackjack): CardNumber 를 enum class 인 denomination 으로 변경

y2gcoder committed:
refactor(blackjack): DefaultBlackJackCardSumCalculator의 sum() 내부 로직 리팩토링: 가능한 조합 중 21과 가까운 합을 구하는 방식으로 변경

y2gcoder committed:
refactor(blackjack): Dealer 의 히트/스테이 여부를 따지는 메서드를 isUnderOver() 로 변경

y2gcoder committed:
feature(blackjack): 상태 패턴 도입을 위한 일급 컬렉션 Cards 도입

Hand를 대체할 예정

y2gcoder committed:
feature(blackjack): Cards 에 isBust() 추가

일급컬렉션에 버스트했는지 여부를 확인할 수 있는 메서드 추가

y2gcoder committed:
feature(blackjack): Cards 에 isBlackjack() 추가

일급컬렉션에 블랙잭인지 여부를 알 수 있는 메서드 추가

y2gcoder committed:
feature(blackjack): 블랙잭 각 상태(Hit, Bust, Stay, Blackjack) 추가

상태 패턴을 사용하기 위해 각 상태를 추가. 각 참가자들은 해당 상태를 보유 및 변환할 수 있을듯

y2gcoder committed:
feature(blackjack): 블랙잭 각 상태(Hit, Bust, Stay, Blackjack)에 profit을 추가

Hit: illegalStateException 추가
Bust: 무조건 베팅금액을 잃기 때문에 -베팅금액
Stay: 기본적으로 승리하면 베팅금액만큼 수익을 얻기 때문에 +배팅금액
Blackjack: 블랙잭으로 승리하면 베팅금액 * 1.5 수익을 얻기 때문에 +(베팅금액 * 1.5)

y2gcoder committed:
refactor(blackjack): ParticipantAccount2 -> ParticipantAccount 로 변경

리팩토링 과정에서 놓친 이름 변경 반영

y2gcoder committed:
feature(blackjack): State 를 사용하는 Participant2, Player2 추가

2단계 블랙잭 구현까지 반영

y2gcoder committed:
feature(blackjack): 3단계 - 블랙잭(딜러) 구현

상태 기반 구현 완료.
플레이어 승리 평가하는 BlackjackGameJudgeStrategy 인터페이스를 Dealer2에 주입하여 위임
DefaultBlackjackGameJudgeStrategy 구현체는 책임 사슬 패턴으로 평가 로직을 리팩토링

y2gcoder committed:
feature(blackjack): 4단계 - 블랙잭(베팅) 기능 구현

기존 블랙잭 기능 완전 대체 완료
dealer 에게 플레이어와의 승부 결과 및 수익률 계산 책임(위임 사용해서 실제 책임은 분산)
List<Money>에 대한 확장함수 추가

y2gcoder committed:
refactor(blackjack): BlackjackGame 에서 콘솔 입출력 UI 완전 분리

콘솔 입력 - ConsoleOutput
콘솔 출력 - ConsoleOutput
로 각각 분리

y2gcoder committed:
refactor(blackjack): BlackjackGame 플로우 처리를 BlackjackService로 위임

BlackjackService 는 참가자 초기화, 플레이어 베팅, 플레이어 턴 진행, 딜러 턴 진행, 참가자 수익 계산 들을 실행

y2gcoder commented:
안녕하십니까, 한솔님!

한솔님 피드백을 받고 지금 구조로는 답이 없겠다 싶어서 

싹다 고쳤습니다!

이번에는 말씀해주셨던 

- 메서드 추상화 레벨
- 컨트롤러와 서비스 로직의 분리
- 딜러가 수익 계산 처리

등을 고려해서 다시 짜보았습니다!

잘 부탁드립니다!!

catsbi commented on src/main/kotlin/blackjack/BlackjackService.kt:
```diff
@@ -0,0 +1,126 @@
+
+import blackjack.ui.DrawAnswer
+
+class BlackjackService(private val deck: Deck) {
```
service 계층에서 상태(state)를 가진다는건 재사용성이 떨어진다는 말과 동일할 것 같아요. deck은 인스턴스변수가 아닌 함수의 매개변수로도 표현되기 충분할 것 같네요!

catsbi commented on src/main/kotlin/blackjack/BlackjackService.kt:
```diff
@@ -0,0 +1,126 @@
+        return dealer to players
+    }
+
+    private fun initializeDealer() = Dealer(List(2) { deck.draw() })
```
List(2) { deck.draw() } 이 행위가 딜러와 플레이어 모두 동일한 행위로 보이네요. 각각 중복된 코드는 줄여볼 수 있을 것 같아요.

catsbi commented on src/main/kotlin/blackjack/BlackjackService.kt:
```diff
@@ -0,0 +1,126 @@
+        participant.stay()
+    }
+
+    fun calculateAllPlayerProfits(
```
게임 서비스에서 수익율 계산까지 같이하는것이 SRP를 위반한다는 생각도 좀 있네요. ProfitCalculateService등으로 책임 위임을 해보는것도 괜찮을 것 같아요.

catsbi commented on src/main/kotlin/blackjack/BlackjackService.kt:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    private fun handlePlayerTurn(
```
딜러와 플레이어가 turn이라는 행위는 공통인데, 실제 내부 동작은 다르다면 이 역시 전략패턴등으로 분리해볼 수 있지 않을까요?

catsbi commented on src/main/kotlin/blackjack/Money.kt:
```diff
@@ -0,0 +1,36 @@
+package blackjack
+
+@JvmInline
+value class Money(private val amount: Double) : Comparable<Money> {
```
Money 값 객체를 잘 만들어주셨네요 👍

catsbi commented on src/main/kotlin/blackjack/ui/BlackjackGame.kt:
```diff
@@ -0,0 +1,55 @@
+import blackjack.ui.ConsoleOutput.Companion.revealAllParticipantsHeldCardsAndScores
+import blackjack.ui.ConsoleOutput.Companion.revealParticipantHeldCards
+
+fun main() {
```
UI로직과 비즈니스 로직이 혼용되어 있는 것 같아요. 좀 더 분리를 고민해보면 어떨까요? UI계층과 뒤섞인 비즈니스로직은 테스트를 작성하기 까다롭거든요 😓

catsbi commented on src/main/kotlin/blackjack/PlayerResult.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack
+
+enum class PlayerResult {
```
이 정보들은 sealed class를 이용해봐도 좋을 것 같네요.

catsbi commented on src/main/kotlin/blackjack/Running.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack
```
현재 패키지가 blackjack 하나에 모두 있다보니 참조의 지역성 측면에선 유리할수 있어도 가독성 측면에서는 좀 떨어지는 것 같네요. 

service나 domain, dto 등의 적절한 레이어 경계를 패키지로 좀 더 해보는것을 권장드려요.

catsbi reviewed (approved):
영근님, 4단계 피드백 반영까지 잘 해주셨어요. 
슬랙 알람이 온 줄 몰라서 리뷰를 늦게 해드렸네요 죄송합니다 😢 

피드백들을 반영하기위해 싹 갈아엎는 작업을 해주셨는데, 고생은 했을지라도 그만큼 인사이트를 얻으셨기를 바랍니다. 

몇 가지 추가 피드백 남겨드렸는데, 미션 요구사항은 구현되었다고 보여 머지하도록 하겠습니다. 
블랙잭 미션 진행하느라 고생 많으셨어요 👍
